wave_containers/include/wave/containers/measurement_container.hpp

```diff
@@ -0,0 +1,313 @@
+    // Specify that the time key corresponds to the time_point member
+    struct time_key : member<T, TimeType, &T::time_point> {};
+    // Specify that the sensor key corresponds to the sensor_id member
+    struct sensor_key : member<T, decltype(T::sensor_id), &T::sensor_id> {};
+
+    // Define a composite key which combines the above two keys
+    // This lets us search elements sorted by sensor id first, then time
+    struct sensor_and_time_key : composite_key<T, sensor_key, time_key> {};
+
+    // These types are used as tags to retrieve each index after the
+    // multi_index_container is generated
+    struct time_index {};
+    struct sensor_index {};
+    struct composite_index {};
+
+    // Define an index for each key. Each index will be accessible via its tag
+    struct indices : indexed_by<
+            ordered_non_unique<tag<time_index>, time_key>,
+            ordered_non_unique<tag<sensor_index>, sensor_key>,
+            ordered_unique<tag<composite_index>, sensor_and_time_key>> {
+    };
+
+    // Note we use separate struct definitions above, instead of typedefs, to
+    // reduce the length of the name printed by the compiler.
+
+    // Finally, define the multi_index_container type.
+    // This is the container type which can actually be used to make objects
+    using type = boost::multi_index_container<T, indices, std::allocator<T>>;
+
+    // For convenience, get the type of the composite index, using its tag
+    using composite_type = typename type::template index<composite_index>::type;
+};
+}  // namespace internal
+
+/** Container which stores and transparently interpolates measurements.
+ *
+ * @tparam T is the stored measurement type. The Measurement class template
+ * in wave/containers/measurement.hpp is designed to be used here.
+ *
+ * However, any class can be used that has the following public members:
+ *   \li \c  time_point (of type \c wave::TimeType)
+ *   \li \c sensor_id (any type sortable by \c std::less)
+ *   \li \c value (any type)
+ *
+ * Additionally, the non-member function
+ *   \code{.cpp}
+ *   interpolate(const T&, const T&, const TimeType&)
+ *   \endcode
+ * must be defined for type \c T.
+ */
+template<typename T>
+class MeasurementContainer {
+ public:
+    // Types
+
+    /** Alias for the template parameter, giving the type of Measurement stored
+     * in this container */
+    using MeasurementType = T;
+    /** Alias for the measurement's value.
+     * Note this does *not* correspond to a typical container's value_type. */
+    using ValueType = decltype(MeasurementType::value);
+    /** Alias for template parameter giving the type of the sensor id */
+    using SensorIdType = decltype(MeasurementType::sensor_id);
+
+    using iterator = typename
+    internal::measurement_container<T>::composite_type::iterator;
+    using const_iterator = typename
+    internal::measurement_container<T>::composite_type::const_iterator;
+    using size_type = std::size_t;
```

Comment:
It is public and can be used from outside the class, just like `std::vector<T>::size_type` (which you should technically use for vector operations instead of `std::size_t`, if you're not using `auto`).

(That it is set to `std::size_t` is a shortcut, yes, but the point is clients don't have to take shortcuts)